### PR TITLE
added movies list page and navigation to movie details page

### PR DESCRIPTION
**Pull request by LautaroSanchezGl**
-Added a scrollable movies list view
-Added navigation back and forth between movies list view and movie details view
-Added functionality to fetch data from local Json and parse it(repository class).

DEMO
[capture2.webm](https://github.com/LautaroSanchezGl/TheMovieDB/assets/142030231/f237a1f3-c50c-46fb-b7d1-f33b760210fb)
[capture3.webm](https://github.com/LautaroSanchezGl/TheMovieDB/assets/142030231/548e1068-104b-48ce-831a-ac84ee48d736)
[capture4.webm](https://github.com/LautaroSanchezGl/TheMovieDB/assets/142030231/5f8b057e-09c1-4049-8876-779f4d153a60)

### CODE REVIEW

File: lib/custom_widgets/popularity_adult_and_original_language.dart [hunk 1]
```diff
@@ -0,0 +1,35 @@
+import 'package:flutter/material.dart';
+import 'middle_text.dart';
+
+class PopularityAdultAndLanguage extends StatelessWidget {
+  const PopularityAdultAndLanguage(
+      {super.key,
+      required this.popularity,
+      required this.adult,
+      required this.originalLanguage});
```
Comment (Kevin-Guillen) [inline lib/custom_widgets/popularity_adult_and_original_language.dart]: trailing comma

File: lib/custom_widgets/popularity_adult_and_original_language.dart [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+  static const String language = 'Language:';
+  static const String staticPopularity = 'Popularity:';
+  static const String staticAdult = 'Adult:';
```
Comment (Kevin-Guillen) [inline lib/custom_widgets/popularity_adult_and_original_language.dart]: i would remove 'static' from the const name. Maybe call it something like pupularityLabel?

File: lib/views/movie_list_view.dart [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+            },
+            separatorBuilder: (BuildContext context, int index) {
+              return const Divider(
+                height: dividerHeight,
+              );
+            },
+          );
+        }
+        return const Text(
+          'An error has occurred while loading data',
```
Comment (Kevin-Guillen) [inline lib/views/movie_list_view.dart]: Constant

File: lib/custom_widgets/popularity_adult_and_original_language.dart [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+  final num popularity;
```
Comment (Kevin-Guillen) [inline lib/custom_widgets/popularity_adult_and_original_language.dart]: in relation to the previous comment , i would rename this pupularityValue

**Review by FranciscoGaraventa — CHANGES_REQUESTED** [review]
You can avoid using this asset file for images; you can use posterPath directly instead. Using this base url combined with the posterPath parameter
'https://image.tmdb.org/t/p/w500/$posterPath'

File: lib/entities/movie.dart [hunk 5]
```diff
@@ -1,26 +1,68 @@
...
+        releaseDate = json['release_date'],
+        voteAverage = json['vote_average'],
+        overview = json['overview'] {
+    for (var id in json['genre_ids']) {
+      for (var genreItem in genresList) {
+        if (genreItem['id'] == id) {
+          genres.add(genreItem['name']);
+        }
+      }
+    }
```
Comment (FranciscoGaraventa) [inline lib/entities/movie.dart]: Instead of early mapping the genres by name, you could leave it directly as they come with the ids, then in the UI where you actually need to see the names there if you can do this kind of mapping where you get the name by id

File: lib/repositories/movie_repository.dart [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+import '../entities/movie.dart';
+
+class MovieRepository {
+  final GenreRepository genreRepository = GenreRepository();
+
+  Future<List<Movie>> getMovies() async {
+    final String response = await rootBundle.loadString('assets/movies.json');
+    List<dynamic> data = List.from(json.decode(response));
+    List<dynamic> genres = await genreRepository.getGenres();
+    return data.map((movie) => Movie.fromJson(movie, genres)).toList();
```
Comment (FranciscoGaraventa) [inline lib/repositories/movie_repository.dart]: trailing comma

File: lib/views/movie_details_view.dart [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+import '../custom_widgets/genres_row.dart';
+import '../custom_widgets/movie_backdrop_and_title_stack.dart';
+import '../custom_widgets/poster_and_overview_stack.dart';
+import '../custom_widgets/release_date_and_vote_average_row.dart';
+import '../entities/movie.dart';
+import '../ui_constants.dart';
+
+class MovieDetailsVIew extends StatelessWidget {
+  const MovieDetailsVIew({super.key, required this.movie});
```
Comment (FranciscoGaraventa) [inline lib/views/movie_details_view.dart]: trailing comma

File: lib/views/movie_list_view.dart [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+class _MovieListViewState extends State<MovieListView> {
+  final MovieRepository movieRepository = MovieRepository();
+  static const double imageHeight = 400;
+  static const double dividerHeight = 20;
+
+  @override
+  Widget build(BuildContext context) {
+    return FutureBuilder(
+      builder: (context, AsyncSnapshot snapshot) {
```
Comment (FranciscoGaraventa) [inline lib/views/movie_list_view.dart]: I recommend to specify the type of the parameter in this case you have BuildContext and AsyncSnapshot<List<Movie>>, this allow you to avoid the casting on line 26

File: pubspec.yaml [hunk 9]
```diff
@@ -52,6 +52,10 @@ dev_dependencies:
 
 # The following section is specific to Flutter packages.
 flutter:
+  assets:
+    - assets/movies.json
+    - assets/genres.json
```
Comment (FranciscoGaraventa) [inline pubspec.yaml]: This could be inside a mock folder

File: lib/views/movie_list_view.dart (hunk 3 above)
Comment (LautaroSanchezGl) [reply lib/views/movie_list_view.dart]: done

File: lib/custom_widgets/popularity_adult_and_original_language.dart (hunk 2 above)
Comment (LautaroSanchezGl) [reply lib/custom_widgets/popularity_adult_and_original_language.dart]: done

File: lib/custom_widgets/popularity_adult_and_original_language.dart (hunk 1 above)
Comment (LautaroSanchezGl) [reply lib/custom_widgets/popularity_adult_and_original_language.dart]: done

File: lib/custom_widgets/popularity_adult_and_original_language.dart (hunk 4 above)
Comment (LautaroSanchezGl) [reply lib/custom_widgets/popularity_adult_and_original_language.dart]: done

File: lib/entities/movie.dart (hunk 5 above)
Comment (LautaroSanchezGl) [reply lib/entities/movie.dart]: done

File: lib/repositories/movie_repository.dart (hunk 6 above)
Comment (LautaroSanchezGl) [reply lib/repositories/movie_repository.dart]: done

File: lib/views/movie_details_view.dart (hunk 7 above)
Comment (LautaroSanchezGl) [reply lib/views/movie_details_view.dart]: done

File: lib/views/movie_list_view.dart (hunk 8 above)
Comment (LautaroSanchezGl) [reply lib/views/movie_list_view.dart]: done

File: pubspec.yaml (hunk 9 above)
Comment (LautaroSanchezGl) [reply pubspec.yaml]: done

File: lib/repositories/genre_repository.dart [hunk 10]
```diff
@@ -0,0 +1,16 @@
+import 'dart:convert';
+
+import 'package:flutter/services.dart';
+
+class GenreRepository {
+  Future<List<dynamic>> getGenreNamesById(List<num> ids) async {
```
Comment (matiasilveiro-gl) [inline lib/repositories/genre_repository.dart]: Should this be a list of `dynamic` type? Hmmm dangerous...

File: lib/ui_constants.dart [hunk 11]
```diff
@@ -15,4 +15,5 @@ abstract class MovieDetailsUiConstants {
   static const double overviewPadding = 150.0;
   static const double borderRadius = 30;
   static const String releaseDateTextConstant = 'Available on';
+  static const String imagesRouteBeginning = 'https://image.tmdb.org/t/p/w500/';
```
Comment (matiasilveiro-gl) [inline lib/ui_constants.dart]: Just to follow the convention, I'd rename this as `imageBaseUrl`, but it's your choice!

File: lib/views/movie_list_view.dart [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+                height: dividerHeight,
+              );
+            },
+          );
+        }
+        return const Text(
+          errorMessage,
+        );
+      },
+      future: movieRepository.getMovies(),
```
Comment (matiasilveiro-gl) [inline lib/views/movie_list_view.dart]: We can add some Flutter best practices here. Please read this:
https://api.flutter.dev/flutter/widgets/FutureBuilder-class.html

> If the future is created at the same time as the FutureBuilder, then every time the FutureBuilder's parent is rebuilt, the asynchronous task will be restarted. A general guideline is to assume that every build method could get called every frame, and to treat omitted calls as an optimization.

TLDR, you could make this widget Stateful, declare a Future<T> movies at the beginning, init that future in the initState method, and use that future in the FutureBuilder. Like here:
https://stackoverflow.com/a/68966640

File: lib/custom_widgets/genres_row.dart [hunk 13]
```diff
@@ -1,23 +1,43 @@
...
+                    genre: genre,
+                  ),
+                )
+                .toList(),
+          );
+        } else {
+          return const Text(errorMessage);
+        }
+      },
+      future: genreRepository.getGenreNamesById(
```
Comment (matiasilveiro-gl) [inline lib/custom_widgets/genres_row.dart]: Same as above

File: lib/repositories/genre_repository.dart (hunk 10 above)
Comment (LautaroSanchezGl) [reply lib/repositories/genre_repository.dart]: what is life without a little risk???? you wanna know?... it's.........boooring!! jaja
done

File: lib/ui_constants.dart (hunk 11 above)
Comment (LautaroSanchezGl) [reply lib/ui_constants.dart]: done

File: lib/views/movie_list_view.dart (hunk 12 above)
Comment (LautaroSanchezGl) [reply lib/views/movie_list_view.dart]: i took my time and i read it! and it's done.

File: lib/custom_widgets/genres_row.dart (hunk 13 above)
Comment (LautaroSanchezGl) [reply lib/custom_widgets/genres_row.dart]: done

File: lib/custom_widgets/genres_row.dart
Comment (matiasilveiro-gl) [inline lib/custom_widgets/genres_row.dart:1]: You could add a loading widget when the `snapshot.connectionState == ConnectionState.loading`.
Also, to improve readability you could place the `future` parameter on top of the `builder` in the `FutureBuilder`.

File: lib/repositories/movie_repository.dart [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+  Future<List<Movie>> getMovies() async {
+    final String response =
+        await rootBundle.loadString('assets/mock_data/movies.json');
+    List<dynamic> data = List.from(json.decode(response));
+    return data
+        .map(
+          (movie) => Movie.fromJson(
+            movie,
+          ),
+        )
```
Comment (matiasilveiro-gl) [inline lib/repositories/movie_repository.dart]: In this case, I believe that these trailing commas are not needed

File: lib/views/movie_list_view.dart [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+                    Padding(
+                      padding: const EdgeInsets.all(
+                          MovieDetailsUiConstants.releaseDateRowPadding),
+                      child: Image.network(
+                        '${MovieDetailsUiConstants.imageBaseUrl}${movies[index].posterPath}',
+                      ),
+                    ),
+                  ],
+                ),
+              );
```
Comment (matiasilveiro-gl) [inline lib/views/movie_list_view.dart]: You could extract this as a standalone widget, improving readability

File: lib/views/movie_list_view.dart
Comment (matiasilveiro-gl) [inline lib/views/movie_list_view.dart:1]: Same as above

File: lib/custom_widgets/genres_row.dart
Comment (LautaroSanchezGl) [reply lib/custom_widgets/genres_row.dart:1]: done

File: lib/repositories/movie_repository.dart (hunk 14 above)
Comment (LautaroSanchezGl) [reply lib/repositories/movie_repository.dart]: done

File: lib/views/movie_list_view.dart (hunk 15 above)
Comment (LautaroSanchezGl) [reply lib/views/movie_list_view.dart]: done

File: lib/views/movie_list_view.dart
Comment (LautaroSanchezGl) [reply lib/views/movie_list_view.dart:1]: done

File: lib/custom_widgets/movie_gesture_detector.dart [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+  final Movie movie;
+
+  @override
+  Widget build(BuildContext context) {
+    return GestureDetector(
+      onTap: () {
+        Navigator.push(
+          context,
+          MaterialPageRoute(
+            builder: (context) => MovieDetailsVIew(
```
Comment (Kevin-Guillen) [inline lib/custom_widgets/movie_gesture_detector.dart]: little mistake with capital letter on the class name

File: lib/custom_widgets/movie_gesture_detector.dart [hunk 17]
```diff
@@ -0,0 +1,48 @@
+import 'package:flutter/material.dart';
+
+import '../entities/movie.dart';
+import '../text_styles/custom_text_style.dart';
+import '../ui_constants.dart';
+import '../views/movie_details_view.dart';
+
+class MovieGestureDetector extends StatelessWidget {
```
Comment (Kevin-Guillen) [inline lib/custom_widgets/movie_gesture_detector.dart]: i think we could find another name for this class , something more related to what the widget contains , not its functionality. Something like MovieInfo maybe?

File: lib/repositories/genre_repository.dart [hunk 18]
```diff
@@ -0,0 +1,17 @@
+import 'dart:convert';
+
+import 'package:flutter/services.dart';
+
+class GenreRepository {
+  Future<List<String>> getGenreNamesById(List<num> ids) async {
+    final String response =
+        await rootBundle.loadString('assets/mock_data/genres.json');
+    List<dynamic> genres = json.decode(response)['genres'];
+    return genres
```
Comment (FranciscoGaraventa) [inline lib/repositories/genre_repository.dart:11]: Instead to transform the genres to a `List<String>` use it as a `List<Genres>`

File: lib/views/movie_details_view.dart [hunk 19]
```diff
@@ -0,0 +1,67 @@
...
+                PopularityAdultAndLanguage(
+                  popularityValue: movie.popularity,
+                  adult: movie.adult,
+                  originalLanguage: movie.originalLanguage,
+                ),
+                const SizedBox(
+                  height: MovieDetailsUiConstants.sizedBoxHeight,
+                ),
+                Genres(
+                  genres: movie.genres,
```
Comment (FranciscoGaraventa) [inline lib/views/movie_details_view.dart:51]: Here you need to transform the genres `(List<int>)` in a `List<Genres>` to use it on Genre widget

File: lib/custom_widgets/movie_gesture_detector.dart (hunk 16 above)
Comment (LautaroSanchezGl) [reply lib/custom_widgets/movie_gesture_detector.dart]: thanks!! done!

File: lib/custom_widgets/movie_gesture_detector.dart (hunk 17 above)
Comment (LautaroSanchezGl) [reply lib/custom_widgets/movie_gesture_detector.dart]: done!!

File: lib/repositories/genre_repository.dart (hunk 18 above)
Comment (LautaroSanchezGl) [reply lib/repositories/genre_repository.dart:11]: done

File: lib/views/movie_details_view.dart (hunk 19 above)
Comment (LautaroSanchezGl) [reply lib/views/movie_details_view.dart:51]: done

File: lib/entities/genre.dart [hunk 20]
```diff
@@ -0,0 +1,18 @@
...
+    required this.name,
+  });
+
+  factory Genre.fromJson(Map<String, dynamic> json) {
+    final id = json['id'];
+    final name = json['name'];
+    return Genre(
+      id: id,
+      name: name,
+    );
```
Comment (matiasilveiro-gl) [inline lib/entities/genre.dart]: Super minor: you could directly
```
return Genre(
      id: json['id'],
      name: json['name'],
    );
```

File: lib/views/movie_list_view.dart [hunk 21]
```diff
@@ -0,0 +1,60 @@
...
+        if (snapshot.connectionState == ConnectionState.waiting) {
+          return const Center(
+            child: CircularProgressIndicator(
+              color: Colors.white,
+            ),
+          );
+        } else if (snapshot.hasData) {
+          List<Movie>? movies = snapshot.data;
+          return ListView.separated(
+            itemCount: movies!.length,
```
Comment (matiasilveiro-gl) [inline lib/views/movie_list_view.dart]: I wouldn't force a null-unwrapping here. I mean, we all know that if `snapshot.hasData`, `snapshot.data` won't be null. 
But just for the sake of using Dart's null safety features, I'd make this change:
`List<Movie> movies = snapshot.data ?? [];`

File: lib/custom_widgets/genres_row.dart [hunk 22]
```diff
@@ -1,23 +1,63 @@
...
+                  ),
+                )
+                .toList(),
+          );
+        } else {
+          return const Text(
+            Genres.errorMessage,
+          );
+        }
+      },
```
Comment (FranciscoGaraventa) [inline lib/custom_widgets/genres_row.dart]: You can change this code to something like this:
```
if (snapshot.connectionState == ConnectionState.waiting) {
          return const Center(
            child: CircularProgressIndicator(
              color: Colors.white,
            ),
          );
        } 
if (snapshot.hasData) {
          return Row(
            mainAxisAlignment: MainAxisAlignment.spaceEvenly,
            children: snapshot.data!
                .map(
                  (genre) => RowGenreItem(
                    genre: genre.name,
                  ),
                )
                .toList(),
          );
        } 
          return const Text(
            Genres.errorMessage,
          );
        }
```

File: lib/entities/genre.dart (hunk 20 above)
Comment (FranciscoGaraventa) [inline lib/entities/genre.dart]: Here you can directly made this
```
Genre(
    id: json['id'],
    name: json['name'],
    )
```

File: lib/views/movie_list_view.dart [hunk 23]
```diff
@@ -0,0 +1,60 @@
...
+class MovieListView extends StatefulWidget {
+  const MovieListView({super.key});
+
+  @override
+  State<MovieListView> createState() => _MovieListViewState();
+}
+
+class _MovieListViewState extends State<MovieListView> {
+  final MovieRepository movieRepository = MovieRepository();
+  Future<List<Movie>>? movies;
```
Comment (matiasilveiro-gl) [inline lib/views/movie_list_view.dart]: This should be `Future<List<Movie>> movies`, as that's the return type of the repository method

**Review by matiasilveiro-gl — APPROVED** [review]
LGTM, with minor comments

File: lib/custom_widgets/genres_row.dart (hunk 22 above)
Comment (LautaroSanchezGl) [reply lib/custom_widgets/genres_row.dart]: done

File: lib/entities/genre.dart (hunk 20 above)
Comment (LautaroSanchezGl) [reply lib/entities/genre.dart]: done

Comment (LautaroSanchezGl) [reply lib/entities/genre.dart]: done

File: lib/views/movie_list_view.dart (hunk 21 above)
Comment (LautaroSanchezGl) [reply lib/views/movie_list_view.dart]: jajaj, fran made a comment related to this above, and suggested something similar but not exactly the same... so, to make you both happy, i implemented this solution with movies snapshot data, and did fran solution to genres snapshot data jaaja.... hope you don't mind!! jaajaj

File: lib/views/movie_list_view.dart (hunk 23 above)
Comment (LautaroSanchezGl) [reply lib/views/movie_list_view.dart]: done

**Review by FranciscoGaraventa — APPROVED** [review]
LG

**Review by matiasilveiro-gl — APPROVED** [review]

**Review by Kevin-Guillen — APPROVED** [review]
LGTM
